### PR TITLE
feat(core): New Redis Cluster, TLS, mTLS and Debug configs for more connection options and better error logging

### PR DESCRIPTION
**Pull request by maspio**
## Summary

Until now n8n only had very limited support to secure Redis connections with TLS. We could just set `QUEUE_BULL_REDIS_TLS=true` to enable TLS/SSL, but that only worked if Redis was serving certificates verifiable by known Root CA certificates. Any configuration to support self-signed certificates was missing. Also minimal to no errors were logged, when connections to Redis failed, making deployment issues very frustrating. Enterprises expect to use existing TLS and mTLS services with self-signed certificates and often backed down from deploying n8n in critical infrastructure without encryption options.

After a lot of trial and error with different Redis TLS deployments a set of improvements was developed to improve support for most used TLS and mTLS service configurations. Only minor refactoring were done, e.g. to improve logging errors. All existing configuration options are still working with new config options to support TLS scenarios like:
* only a server certificate but no client certificate
* both client and server have a certificate (mtls), either the same or different ones
* support certificates with or without passphrase protection
* different hostname/ip/ip-family combinations and self-signed certificates

New [Redis TLS config](https://github.com/n8n-io/n8n/blob/3558fd1e1ee87d6e02f9a3a61d86917818b42f1e/packages/%40n8n/config/src/configs/scaling-mode.config.ts#L23) options:
* `QUEUE_BULL_REDIS_TLS_CA_FILE` - Root CA certificate file path
* `QUEUE_BULL_REDIS_TLS_CERT_FILE` - Client certificate file path
* `QUEUE_BULL_REDIS_TLS_CERT_KEY_FILE` - The client certificate's private key.
* `QUEUE_BULL_REDIS_TLS_CERT_KEY_PASS` - The passphrase to decrypt the client certificate's private key
* `QUEUE_BULL_REDIS_TLS_SERVERNAME` - SNI extension servername for TLS handshake sometimes required to solve hostname TLS resolution issues.
* `QUEUE_BULL_REDIS_TLS_REJECT_UNAUTHORIZED` - If unauthorized SSL connections should be rejected. If nothing helps just ignore the problem for a while :D
* `QUEUE_BULL_REDIS_IPV6` - Whether to use only IPv6 to avoid ENOTFOUND errors when a hostname only resolves to an IPv6 address. Default uses IPv4 only. Backwards compatible with existing `QUEUE_BULL_REDIS_DUALSTACK` env var which enables support for both IPv4 and IPv6. All possible scenarios now supported.
* `QUEUE_BULL_REDIS_DNS_LOOKUP_STRATEGY` - DNS lookups can be error prone, especially in combination with TLS certificates. Especially AWS Elasticache cluster connections often lead to invalid certificate errors due to hostname/ip mismatches.

Better Debugging, connection logging and options for TLS related connection issues. See [redis-client.service.ts](https://github.com/n8n-io/n8n/blob/3558fd1e1ee87d6e02f9a3a61d86917818b42f1e/packages/cli/src/services/redis-client.service.ts#L253):
* [QUEUE_BULL_REDIS_DEBUG](https://github.com/n8n-io/n8n/blob/3558fd1e1ee87d6e02f9a3a61d86917818b42f1e/packages/%40n8n/config/src/configs/scaling-mode.config.ts#L83C8-L83C30) - Enable verbose ioredis client logging to debug connection and command issues. See [enableRedisDebug()](https://github.com/n8n-io/n8n/blob/3558fd1e1ee87d6e02f9a3a61d86917818b42f1e/packages/cli/src/services/redis-client.service.ts#L33)
* Improve initial connection error logging. With [lazyConnect=true](https://github.com/n8n-io/n8n/blob/3558fd1e1ee87d6e02f9a3a61d86917818b42f1e/packages/cli/src/services/redis-client.service.ts#L194) ioredis won't auto connect on client creation, but later when the first command is executed. A unified [connect()](https://github.com/n8n-io/n8n/blob/3558fd1e1ee87d6e02f9a3a61d86917818b42f1e/packages/cli/src/services/redis-client.service.ts#L265C10-L265C17) function for both regular & cluster client is [called once after client setup is done](https://github.com/n8n-io/n8n/blob/3558fd1e1ee87d6e02f9a3a61d86917818b42f1e/packages/cli/src/services/redis-client.service.ts#L100) to log errors that ioredis is hiding when doing auto connect with retry strategies.
* [onReconnectError](https://github.com/n8n-io/n8n/blob/3558fd1e1ee87d6e02f9a3a61d86917818b42f1e/packages/cli/src/services/redis-client.service.ts#L287) logging to surface helpful network or TLS errors
* [getHostnameResolutionIpAddressFamily()](https://github.com/n8n-io/n8n/blob/3558fd1e1ee87d6e02f9a3a61d86917818b42f1e/packages/cli/src/services/redis-client.service.ts#L163C9-L163C45) helper to support all IPv4/IPv6 DNS lookup variants
* [getDnsLookupFunction()](https://github.com/n8n-io/n8n/blob/4bf0ca135bb341f4dc586212100f0de59de02228/packages/cli/src/services/redis-client.service.ts#L257) to support different IP/Hostname resolution strategies.
* [readTlsFileSync()](https://github.com/n8n-io/n8n/blob/3558fd1e1ee87d6e02f9a3a61d86917818b42f1e/packages/cli/src/services/redis-client.service.ts#L234C10-L234C25) util to read TLS certificate files synchronously once on client creation with graceful error handling. 
* Setting redis client [connectionName](https://github.com/n8n-io/n8n/blob/a23f676161f37eafb45a8b2c396bef54ba926bc3/packages/cli/src/services/redis-client.service.ts#L182) to a unique identifier per connected n8n instances to identify the Redis commands they send on they server.

Not included in this PR:
* NAT Mapping support - Sometimes a Redis cluster is hosted within a internal network that can only be accessed via a NAT (Network Address Translation). NAT configuration are often required, e.g. to access ElastiCache from outside AWS. [See IoRedis NAT Mapping ](https://github.com/redis/ioredis?tab=readme-ov-file#nat-mapping)
* Redis Sentinel Support - [See ioredis sentinel](https://github.com/redis/ioredis?tab=readme-ov-file#sentinel)

Docker Image based on this PR for testing before officially released
```bash
docker pull ghcr.io/n8n-io/n8n:branch-more-scaling-mode-queue-bull-redis-tls-configs
```

## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

File: packages/@n8n/config/src/configs/scaling-mode.config.ts [hunk 1]
```diff
@@ -53,6 +53,18 @@ class RedisConfig {
 	@Env('QUEUE_BULL_REDIS_TLS')
 	tls: boolean = false;
 
+	/** SSL certificate authority */
+	@Env('QUEUE_BULL_REDIS_TLS_CA')
```
Comment (Joffcom) [inline packages/@n8n/config/src/configs/scaling-mode.config.ts]: Is this after a file path or the contents? If it is the path to a CA cert maybe the name can be clearer?

Comment (maspio) [reply packages/@n8n/config/src/configs/scaling-mode.config.ts]: Yeah thanks @Joffcom, also thought that. But still in draft mode to figure things out. Also sent out the PR docker image to a customer who has issues connecting to Redis because they enforce SNI servername.

**Comment by maspio** [comment]
@claude review

**Comment by PaulsBecks** [comment]
Hey @maspio. This PR is still very important for us. How can we progress here? Any chance to bring it into the next release?

**Comment by maspio** [comment]
@claude review

**Comment by maspio** [comment]
@claude review

**Comment by marty-sullivan** [comment]
@maspio This PR was brought to my attention by the recent mention to my PR for AWS ElastiCache Serverless above. This is looking really good.

If this is merged, will setting `QUEUE_BULL_REDIS_DNS_LOOKUP_STRATEGY=NONE` effectively set the following required code for ElastiCache Serverless to work?

```typescript
clusterOptions.dnsLookup = (address, callback) => callback(null, address);
``` 

If so, I'd probably lean toward this PR as it is more generic and less focused on ElastiCache. 

That said, I've also found that with ElastiCache Serverless, I need to tweak the following to fix some occasional errors with the integration:

```typescript
clusterOptions.slotsRefreshTimeout = 5000;
clusterOptions.slotsRefreshInterval = 10000;
```

I don't think there are existing flags to set these options. If that is the case, could you also add flags to expose these parameters?

**Comment by maspio** [comment]
@marty-sullivan thanks for getting involved. I will hand this over for review by Monday so hopefully we can release soon.

> If this is merged, will setting QUEUE_BULL_REDIS_DNS_LOOKUP_STRATEGY=NONE effectively set the following required code for ElastiCache Serverless to work?
> ```ts
> clusterOptions.dnsLookup = (address, callback) => callback(null, address);
> ```

Yes, see [getDnsLookupFunction()](https://github.com/n8n-io/n8n/blob/4bf0ca135bb341f4dc586212100f0de59de02228/packages/cli/src/services/redis-client.service.ts#L257)

> ```ts
> clusterOptions.slotsRefreshTimeout = 5000;
> clusterOptions.slotsRefreshInterval = 10000;
>``` 
>[...] could you also add flags to expose these parameters?

Yes, easy to add pure configuration. Thanks for sharing your experience. It's hard to figure out all the edge cases.

**Comment by marty-sullivan** [comment]
@maspio Awesome! Once you've finalized the PR, let me know and I'll be happy to test things in our self-hosted environment
